packages/plugin/src/controllers/api/fields/GroupsController.php

```diff
@@ -0,0 +1,86 @@
+<?php
+
+namespace Solspace\Freeform\controllers\api\fields;
+
+use Solspace\Freeform\Bundles\Fields\Types\FieldTypesProvider;
+use Solspace\Freeform\controllers\BaseApiController;
+use Solspace\Freeform\Freeform;
+use Solspace\Freeform\Records\FieldTypeGroupRecord;
+
+class GroupsController extends BaseApiController
+{
+    private FieldTypesProvider $fieldTypesProvider;
+
+    public function __construct($id, $module, $config = [], FieldTypesProvider $fieldTypesProvider)
+    {
+        parent::__construct($id, $module, $config);
+        $this->fieldTypesProvider = $fieldTypesProvider;
+    }
+
+    protected function get(): object
+    {
+        $freeform = Freeform::getInstance();
+        $groups = FieldTypeGroupRecord::find()->all();
+        $types = $this->fieldTypesProvider->getRegisteredTypes();
+
+        $response = (object) [
+            'types' => [],
+            'groups' => [
+                'hidden' => [],
+                'grouped' => [],
+            ],
+        ];
+
+        if ($freeform->isPro()) {
+            $hiddenFieldTypes = $freeform->settings->getSettingsModel()->hiddenFieldTypes;
+            $unassignedTypes = array_diff($types, array_merge(...array_column($groups, 'types')), $hiddenFieldTypes);
```

Comment:
If you're doing some dark magic kung fu, then please try to make it as readable as possible.
A good thing to do is use newlines and indentation to make it more readable.
The best is using variables and good naming.

```suggestion
            $unassignedTypes = array_diff(
                $types,
                array_merge(
                    ...array_column($groups, 'types')
                ), 
                $hiddenFieldTypes
            );
```

This is pretty genius though! But takes some time to process.
So this would be a prime candidate for pulling it out into a separate, well named variable.

```php
$flattenedAssignedTypes = array_merge(...array_column($groups, 'types')), 
```


```suggestion
            $unassignedTypes = array_diff(
                $types,
                $flattenedAssignedTypes,
                $hiddenFieldTypes
            );
```